pkg/kv/kvserver/replica_raftstorage.go

```diff
@@ -482,59 +482,64 @@ func (r *Replica) applySnapshot(
-		return errors.Wrapf(err, "unable to write HardState to unreplicated SST writer")
-	}
-	// We've cleared all the raft state above, so we are forced to write the
-	// RaftReplicaID again here.
-	if err := r.raftMu.stateLoader.SetRaftReplicaID(
-		ctx, &unreplicatedSST, r.replicaID); err != nil {
-		return errors.Wrapf(err, "unable to write RaftReplicaID to unreplicated SST writer")
-	}
+	writeUnreplicatedSST := func(ctx context.Context, id storage.FullReplicaID, st *cluster.Settings, meta raftpb.SnapshotMetadata, hs raftpb.HardState, sl *logstore.StateLoader) (*storage.MemFile, bool, error) {
+		unreplicatedSSTFile := &storage.MemFile{}
+		unreplicatedSST := storage.MakeIngestionSSTWriter(
+			ctx, st, unreplicatedSSTFile,
+		)
+		defer unreplicatedSST.Close()
+
+		// Clearing the unreplicated state.
+		//
+		// NB: We do not expect to see range keys in the unreplicated state, so
+		// we don't drop a range tombstone across the range key space.
+		unreplicatedPrefixKey := keys.MakeRangeIDUnreplicatedPrefix(id.RangeID)
+		unreplicatedStart := unreplicatedPrefixKey
+		unreplicatedEnd := unreplicatedPrefixKey.PrefixEnd()
+		if err := unreplicatedSST.ClearRawRange(
+			unreplicatedStart, unreplicatedEnd, true /* pointKeys */, false, /* rangeKeys */
+		); err != nil {
+			return nil, false, errors.Wrapf(err, "error clearing range of unreplicated SST writer")
+		}
 
-	// Update Raft entries.
-	r.store.raftEntryCache.Drop(r.RangeID)
+		// Update HardState.
+		if err := sl.SetHardState(ctx, &unreplicatedSST, hs); err != nil {
+			return nil, false, errors.Wrapf(err, "unable to write HardState to unreplicated SST writer")
+		}
+		// We've cleared all the raft state above, so we are forced to write the
+		// RaftReplicaID again here.
+		if err := sl.SetRaftReplicaID(
+			ctx, &unreplicatedSST, id.ReplicaID); err != nil {
+			return nil, false, errors.Wrapf(err, "unable to write RaftReplicaID to unreplicated SST writer")
+		}
 
-	if err := r.raftMu.stateLoader.SetRaftTruncatedState(
-		ctx, &unreplicatedSST,
-		&roachpb.RaftTruncatedState{
-			Index: nonemptySnap.Metadata.Index,
-			Term:  nonemptySnap.Metadata.Term,
-		},
-	); err != nil {
-		return errors.Wrapf(err, "unable to write TruncatedState to unreplicated SST writer")
-	}
+		if err := sl.SetRaftTruncatedState(
+			ctx, &unreplicatedSST,
+			&roachpb.RaftTruncatedState{
+				Index: meta.Index,
+				Term:  meta.Term,
+			},
+		); err != nil {
+			return nil, false, errors.Wrapf(err, "unable to write TruncatedState to unreplicated SST writer")
+		}
 
-	if err := unreplicatedSST.Finish(); err != nil {
-		return err
+		if err := unreplicatedSST.Finish(); err != nil {
+			return nil, false, err
+		}
+		return unreplicatedSSTFile, unreplicatedSST.DataSize > 0, nil
 	}
-	if unreplicatedSST.DataSize > 0 {
+	unreplicatedSSTFile, nonempty, err := writeUnreplicatedSST(
+		ctx, r.ID(), r.ClusterSettings(), nonemptySnap.Metadata, hs, &r.raftMu.stateLoader.StateLoader,
+	)
```

Comment:
what about checking err != nil?

Reply:
Good catch - that CI didn't catch this is surprising[^1]

[^1]: https://cockroachlabs.slack.com/archives/C023S0V4YEB/p1676290099032059